### ISSUE TITLE
Polyline Support

### ISSUE THREAD

**Issue by hallahan** [mapbox/wagyu#49]
Since the the API and functionality is intended to be similar to the Angus Clipper library, it would be good to be able to clip polylines in addition to polygons. As discussed, this will be difficult to incorporate, but it should be done.

Note the Angus docs showing clipping of a mixture of polygons with polylines:

http://www.angusj.com/delphi/clipper/documentation/Docs/Units/ClipperLib/Types/ClipType.htm

**Comment by flippmoke** [comment]
A basic outline to achieve this would be:

- [ ] Expand and test the existing conversion of lines to local_minimums/bounds `build_edges.hpp` and `local_mimum_util.hpp`
- [ ] Update `topology_correction.hpp` to properly support lines
   - [ ] Skip normal processing of lines in some topology correction methods
   - [ ] Determine how to properly handle situations where a line would touch a polygon at two discrete locations
   - [ ] Clean up possible collinear segments of lines and polygons.
- [ ] Add ability to return only lines in build results
- [ ] Add ability to return mixed lines and polygons in build results
    - [ ] Add mapbox/variant as a requirement
    - [ ] Add useage of mapbox/geometry.hpp generalized

**Comment by flippmoke** [comment]
After a lot of consideration, I think I am at the point where I don't think we should ever support lines as a part of the wagyu algorithm. While the original angus clipper does support this methodology, I feel that it is probably best if we completely removed all code paths where lines could be used in the code. 

The reasoning behind this is that there are **much** better ways of clipping polylines. The other algorithms do not seem to fit well either. I feel like wagyu should strictly focus on polygons as this is how it is designed.  -- I will start referencing this ticket as I slowly remove the portions of wagyu that use polylines/line_strings.

**Comment by springmeyer** [comment]
>  I feel like wagyu should strictly focus on polygons as this is how it is designed.

👍

**Comment by paoletto** [comment]
So what would be the alternative to clip polylines defined as mapbox geometry? like a line_string, for example. The impression is that nothing, inside mapbox/deps can actually do it.

**Comment by flippmoke** [comment]
@paoletto our plans are currently to add this to https://github.com/mapbox/spatial-algorithms

This is currently tracked in https://github.com/mapbox/spatial-algorithms/issues/22 where we will integrate wagyu and boost geometry together into a single API. 

In the short term I would suggest using http://www.boost.org/doc/libs/1_66_0/libs/geometry/doc/html/geometry/reference/algorithms/intersection/intersection_3.html for line_strings/points.

**Comment by paoletto** [comment]
@flippmoke thanks!

**Comment by paoletto** [comment]
@flippmoke tried that, it actually seems that boost::geometry offers the functionality, but is a lot more bugged than expected, and fails way too much to be actually used in practice.
Question: if i try to revert locally  ecb33eb , would that be an OK starting point to implement your first two points in the outline?

**Comment by flippmoke** [comment]
@paoletto what is bugged about boost geometry here?

**Comment by paoletto** [comment]
@flippmoke so i tried to follow the example in https://github.com/springmeyer/boost-geometry-line-clipping-testcase, essentially using int64_t as point member type, and scaling them up (and then down) from the [0, 1] source coordinate interval like i was doing with Clipper.

With this approach it seems that a very large line (say, something that spans more than 0.5 in the source coordinates) gets screwed as soon as the clip polygon is somewhat small.

If you want to reproduce and see it, you could try https://github.com/paoletto/PolylineBench as a sample against a Qt 5.11 patched with https://codereview.qt-project.org/#/c/221553/.
If you only want to see it, there's a screencast.mp4 in the sample repo

**Comment by springmeyer** [comment]
@paoletto what version of boost? That testcase was highlighting a bug that was fixed in https://github.com/boostorg/geometry/pull/329

**Comment by flippmoke** [comment]
@paoletto we currently use boost geometry as part of our processing at Mapbox and do not seem to have issues - https://github.com/mapbox/mapnik-vector-tile/blob/master/src/vector_tile_geometry_clipper.hpp#L136

>Question: if i try to revert locally ecb33eb , would that be an OK starting point to implement your first two points in the outline?

This would be very challenging to do for a variety of reasons but if you are really keen on doing this I can try to help. We decided not to look into this because its really not well suited as a clipping algorithms for line_strings and points. Other methods/algorithms are fasters.

**Comment by paoletto** [comment]
@springmeyer  thanks, i was using 1.57 (the one coming from the mason pkg), I will retry and let you know if the problem is fixed :-)

@flippmoke right, i also experienced poor performance with it. However, it was the most robust i could find (others tried failing here or there). I didn't try boost::geometry though (or maybe i did and it had problem, i can't recall. I might also have been trying an outdated boost version exhibiting the bug @springmeyer  mentioned above).
So if boost fixes the issue, i think i can switch, although this will make the lib grow from 800k to 3.2mb (in debug mode though)

**Comment by springmeyer** [comment]
@paoletto the advantage of boost::geometry for line clipping is preservation of direction. We found with http://www.angusj.com/delphi/clipper.php (the origination of the wagyu line clipping code before removal) was that it did not preserve direction/orientation. Also I've never noticed boost::geometry line clipping as a meaningful bottleneck - always been elsewhere so I think its pretty fast.

**Comment by springmeyer** [comment]
> make the lib grow from 800k to 3.2mb (in debug mode though)

Right, boost will swell debug mode sizes, but should be small in release mode. To tame debug mode sizes try `-gline-tables-only` with `clang++` instead of `-g` (if you happen to be using clang++) (https://github.com/mapbox/cpp/blob/master/glossary.md#profiling-build)

**Comment by paoletto** [comment]
@springmeyer  I might be doing something wrong, but it seems boost 1.66 has the same problem.
I suspected it had to do with converting doubles to and from int64_t, with the same code that however used to work with Clipper. But then i tried to switch to doubles (using raw mercator coordinates), and i had the same problem.
I also checked the clipping region, and even with a simple [(-1,1) (2,1) (2,0) (-1,0) (-1,1)] rectangle, when zooming in enough, lines start to fly around (https://imgur.com/a/OXed7).

Maybe it's the usage of the algos (http://paste.debian.net/1012350/)? I quite dumbly used the example linked above, but perhaps that wasn't supposed to work for large lines?
